cmd/notify-mailer/main.go

```diff
@@ -264,68 +264,100 @@ type addressToRecipientMap map[string][]recipient
 
 // readRecipientsList parses the contents of a recipient list file into a list
 // of `recipient` objects.
-func readRecipientsList(filename string) ([]recipient, error) {
+func readRecipientsList(filename string, delimiter rune) ([]recipient, string, error) {
 	f, err := os.Open(filename)
 	if err != nil {
-		return nil, err
+		return nil, "", err
 	}
 
 	reader := csv.NewReader(f)
+	reader.Comma = delimiter
+
+	// Parse header.
 	record, err := reader.Read()
 	if err != nil {
-		return nil, err
-	}
-
-	if len(record) == 0 {
-		return nil, errors.New("no records in CSV")
+		return nil, "", fmt.Errorf("failed to parse header: %w", err)
 	}
 
 	if record[0] != "id" {
-		return nil, errors.New("first field of CSV input must be \"id\"")
+		return nil, "", errors.New("header must begin with \"id\"")
 	}
 
+	// Collect the names of each header column after `id`.
 	var dataColumns []string
 	for _, v := range record[1:] {
 		dataColumns = append(dataColumns, strings.TrimSpace(v))
+		if len(v) == 0 {
+			return nil, "", errors.New("header contains an empty column")
+		}
+	}
+
+	var recordsWithEmptyColumns []int64
+	var recordsWithDuplicateIDs []int64
+	var probsBuff strings.Builder
+	stringProbs := func() string {
+		if len(recordsWithEmptyColumns) != 0 {
+			fmt.Fprintf(&probsBuff, "ID(s) %v contained empty columns and ",
+				recordsWithEmptyColumns)
+		}
+
+		if len(recordsWithDuplicateIDs) != 0 {
+			fmt.Fprintf(&probsBuff, "ID(s) %v were skipped as duplicates",
+				recordsWithDuplicateIDs)
+		}
+
+		if probsBuff.Len() == 0 {
+			return ""
+		}
+		return strings.TrimSuffix(probsBuff.String(), " and ")
```

Comment:
This is a clever way to solve this problem! Here's the way I would typically do it:

```
var probs []string
if (...) {
  probs = append(probs, fmt.Sprintf(...))
}
if (...) {
  probs = append(probs, fmt.Sprintf(...))
}
return strings.Join(probs, " and ")
```

That accounts for both the empty case and the length==1 case.
```